src/interpreter/interpreter-non-det.ts

```diff
@@ -0,0 +1,502 @@
+/* tslint:disable:max-classes-per-file */
+import * as es from 'estree'
+import * as constants from '../constants'
+import * as errors from '../errors/errors'
+import { RuntimeSourceError } from '../errors/runtimeSourceError'
+import { Context, Environment, Frame, Value } from '../types'
+import { primitive, conditionalExpression, literal } from '../utils/astCreator'
+import { evaluateBinaryExpression, evaluateUnaryExpression } from '../utils/operators'
+import * as rttc from '../utils/rttc'
+import Closure from './closure'
+import { cloneDeep, assignIn } from 'lodash'
+import { CUT } from '../constants'
+
+class ReturnValue {
+  constructor(public value: Value) {}
+}
+
+const createEnvironment = (
+  closure: Closure,
+  args: Value[],
+  callExpression?: es.CallExpression
+): Environment => {
+  const environment: Environment = {
+    name: closure.functionName, // TODO: Change this
+    tail: closure.environment,
+    head: {}
+  }
+  if (callExpression) {
+    environment.callExpression = {
+      ...callExpression,
+      arguments: args.map(primitive)
+    }
+  }
+  closure.node.params.forEach((param, index) => {
+    const ident = param as es.Identifier
+    environment.head[ident.name] = args[index]
+  })
+  return environment
+}
+
+const createBlockEnvironment = (
+  context: Context,
+  name = 'blockEnvironment',
+  head: Frame = {}
+): Environment => {
+  return {
+    name,
+    tail: currentEnvironment(context),
+    head,
+    thisContext: context
+  }
+}
+
+const handleRuntimeError = (context: Context, error: RuntimeSourceError): never => {
+  context.errors.push(error)
+  context.runtime.environments = context.runtime.environments.slice(
+    -context.numberOfOuterEnvironments
+  )
+  throw error
+}
+
+const HOISTED_BUT_NOT_YET_ASSIGNED = Symbol('Used to implement hoisting')
+
+function hoistIdentifier(context: Context, name: string, node: es.Node) {
```

Comment:
Should be called declareIdentifier.
Same for
hoistVariableDeclarations
hoistFunctionsAndVariableDeclarationsIdentifiers